torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1152,9 +1158,48 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
   abort(abortReason);
 }
 
+void dump_debugging_info() {
+  LOG(ERROR) << "Dumped debugging information: " << dump_nccl_trace();
+}
+
+void ProcessGroupNCCL::ncclCommsMonitor() {
+  int hearBeatCounter = -1;
+  while (true) {
+    auto heartbeat = heartbeat_.load();
+    if (heartbeat > hearBeatCounter) {
+      hearBeatCounter = heartbeat;
+      // Sleep for enough time so that heartBeat will at least increase.
+      std::this_thread::sleep_for(
+          std::chrono::milliseconds(kWatchdogThreadSleepMillis * 3));
+    } else {
+      // Timeout case and we will first dump the flight recorder
+      // to std::out, and we use a different detached thread to ensure
+      // monitor thread will never be blocked and to dump debugging info
+      // is best effort.
+      std::thread sideThread(dump_debugging_info);
+      sideThread.detach();
+      break;
+    }
+  }
+
+  // Create a error message reported from MonitorThread, so
+  // we throw exception and make the whole process to be killed.
+  const auto exitMsg = c10::str(
+      "[Rank ", rank_, "] NCCL monitor thread terminated with timeout: ");
+  monitorException_ = std::make_exception_ptr(std::runtime_error(exitMsg));
```

Comment:
@malfet had some comments on our pattern of setting the exception this way in our old watchdog code.  I know you're just copying the same pattern, but, i think its probably not the best pattern so would be good to understand why (if there is a reason) or else just do something simpler.

Reply:
I am more than happy to learn what @malfet might say about this.